File: backend/controllers/therapy_controller.py
from fastapi import APIRouter, HTTPException, Depends, status
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from models.schemas import Therapy, TherapyCreate, TherapyUpdate
from services.auth_service import auth_service
from motor.motor_asyncio import AsyncIOMotorDatabase
from typing import List
from datetime import datetime, timezone
import uuid
import logging

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/therapies", tags=["Therapies"])
security = HTTPBearer(auto_error=False)


async def verify_admin(credentials: HTTPAuthorizationCredentials, db: AsyncIOMotorDatabase) -> dict:
    """Verify admin user from token"""
    if not credentials:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Not authenticated")
    
    payload = auth_service.decode_token(credentials.credentials)
    if not payload or payload.get("type") != "access":
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid token")
    
    user = await db.admin_users.find_one({"id": payload["sub"]}, {"_id": 0})
    if not user or not user.get("is_active", True):
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="User not found or disabled")
    
    return user


def create_therapy_routes(db: AsyncIOMotorDatabase):
    """Create therapy CRUD routes"""
    
    @router.get("/", response_model=List[Therapy])
    async def list_therapies(active_only: bool = False):
        """List all therapies (public endpoint)"""
        query = {"is_active": True} if active_only else {}
        therapies = await db.therapies.find(query, {"_id": 0}).sort("display_order", 1).to_list(100)
        return therapies
    
    @router.get("/{therapy_id}", response_model=Therapy)
    async def get_therapy(therapy_id: str):
        """Get a specific therapy (public endpoint)"""
        therapy = await db.therapies.find_one({"id": therapy_id}, {"_id": 0})
        if not therapy:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Therapy not found")
        return therapy
    
    @router.post("/", response_model=Therapy, status_code=status.HTTP_201_CREATED)
    async def create_therapy(
        therapy_data: TherapyCreate,
        credentials: HTTPAuthorizationCredentials = Depends(security)
    ):
        """Create a new therapy (admin only)"""
        await verify_admin(credentials, db)
        
        therapy_id = str(uuid.uuid4())
        therapy_doc = {
            "id": therapy_id,
            **therapy_data.model_dump(),
            "created_at": datetime.now(timezone.utc).isoformat()
        }
        
        await db.therapies.insert_one(therapy_doc)
        logger.info(f"Created therapy: {therapy_data.name}")
        
        return await db.therapies.find_one({"id": therapy_id}, {"_id": 0})
    
    @router.put("/{therapy_id}", response_model=Therapy)
    async def update_therapy(
        therapy_id: str,
        therapy_data: TherapyUpdate,
        credentials: HTTPAuthorizationCredentials = Depends(security)
    ):
        """Update a therapy (admin only)"""
        await verify_admin(credentials, db)
        
        therapy = await db.therapies.find_one({"id": therapy_id})
        if not therapy:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Therapy not found")
        
        update_data = {k: v for k, v in therapy_data.model_dump().items() if v is not None}
        
        if update_data:
            await db.therapies.update_one({"id": therapy_id}, {"$set": update_data})
        
        return await db.therapies.find_one({"id": therapy_id}, {"_id": 0})
    
    @router.delete("/{therapy_id}", status_code=status.HTTP_204_NO_CONTENT)
    async def delete_therapy(
        therapy_id: str,
        credentials: HTTPAuthorizationCredentials = Depends(security)
    ):
        """Delete a therapy (admin only)"""
        await verify_admin(credentials, db)
        
        therapy = await db.therapies.find_one({"id": therapy_id})
        if not therapy:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Therapy not found")
        
        # Also delete associated prices
        await db.prices.delete_many({"therapy_id": therapy_id})
        await db.therapies.delete_one({"id": therapy_id})
        logger.info(f"Deleted therapy: {therapy_id}")
    
    return router

File: backend/controllers/TherapyController.js
/**
 * Therapy Controller
 * Single Responsibility: CRUD operations for therapies
 */

const { v4: uuidv4 } = require('uuid');

class TherapyController {
  constructor(collections) {
    this.collections = collections;
  }

  // GET /api/therapies (public)
  list = async (req, res) => {
    try {
      const { active_only } = req.query;
      const query = active_only === 'true' ? { is_active: true } : {};

      const therapies = await this.collections.therapies
        .find(query, { projection: { _id: 0 } })
        .sort({ display_order: 1 })
        .toArray();

      res.json({
        success: true,
        therapies
      });
    } catch (error) {
      console.error('List therapies error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to list therapies'
      });
    }
  };

  // GET /api/therapies/:id (public)
  get = async (req, res) => {
    try {
      const { id } = req.params;

      const therapy = await this.collections.therapies.findOne(
        { id },
        { projection: { _id: 0 } }
      );

      if (!therapy) {
        return res.status(404).json({
          success: false,
          message: 'Therapy not found'
        });
      }

      res.json({
        success: true,
        therapy
      });
    } catch (error) {
      console.error('Get therapy error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to get therapy'
      });
    }
  };

  // POST /api/admin/therapies (admin)
  create = async (req, res) => {
    try {
      const {
        name,
        short_description,
        full_description,
        image_url,
        icon,
        display_order,
        is_active,
        coming_soon
      } = req.body;

      if (!name || !short_description) {
        return res.status(400).json({
          success: false,
          message: 'Name and short description are required'
        });
      }

      const therapy = {
        id: uuidv4(),
        name,
        short_description,
        full_description: full_description || '',
        image_url: image_url || '',
        icon: icon || 'Sparkles',
        display_order: display_order || 0,
        is_active: is_active !== false,
        coming_soon: coming_soon === true,
        created_at: new Date().toISOString()
      };

      await this.collections.therapies.insertOne(therapy);

      console.log(`✅ Created therapy: ${name}`);

      res.status(201).json({
        success: true,
        therapy
      });
    } catch (error) {
      console.error('Create therapy error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to create therapy'
      });
    }
  };

  // PUT /api/admin/therapies/:id (admin)
  update = async (req, res) => {
    try {
      const { id } = req.params;
      const updateFields = ['name', 'short_description', 'full_description', 'image_url', 'icon', 'display_order', 'is_active', 'coming_soon'];

      const therapy = await this.collections.therapies.findOne({ id });

      if (!therapy) {
        return res.status(404).json({
          success: false,
          message: 'Therapy not found'
        });
      }

      const updateData = {};
      for (const field of updateFields) {
        if (req.body[field] !== undefined) {
          updateData[field] = req.body[field];
        }
      }

      if (Object.keys(updateData).length > 0) {
        await this.collections.therapies.updateOne(
          { id },
          { $set: updateData }
        );
      }

      const updated = await this.collections.therapies.findOne(
        { id },
        { projection: { _id: 0 } }
      );

      res.json({
        success: true,
        therapy: updated
      });
    } catch (error) {
      console.error('Update therapy error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to update therapy'
      });
    }
  };

  // DELETE /api/admin/therapies/:id (admin)
  delete = async (req, res) => {
    try {
      const { id } = req.params;

      const therapy = await this.collections.therapies.findOne({ id });

      if (!therapy) {
        return res.status(404).json({
          success: false,
          message: 'Therapy not found'
        });
      }

      // Also delete associated prices
      await this.collections.prices.deleteMany({ therapy_id: id });
      await this.collections.therapies.deleteOne({ id });

      console.log(`✅ Deleted therapy: ${id}`);

      res.json({
        success: true,
        message: 'Therapy and associated prices deleted'
      });
    } catch (error) {
      console.error('Delete therapy error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to delete therapy'
      });
    }
  };
}

module.exports = TherapyController;
